TASK: Use proper version checks for contentStream events

bwaidelich opened the pull request:
Replaces some `ExpectedVersion::ANY()` occurrences with proper versions the decisions were based on.

**Note:** With this change there are 8 `ExpectedVersion::ANY()` left in the `WorkspaceCommandHandler` that would require the workspace read model to keep track of the version. There are three more occurrences in `CommandSimulator` and `ContentStreamHandling` that are kept by intention

Related: #5058

bwaidelich committed:
TASK: Use proper version checks where applicable

Replaces some `ExpectedVersion::ANY()` occurrences with proper versions the decisions were based on.

**Note:** With this change there are 8 `ExpectedVersion::ANY()` left in the `WorkspaceCommandHandler` that would require the workspace read model to keep track of the version.
There are three more occurrences in `CommandSimulator` and `ContentStreamHandling` that are kept by intention

Related: #5058

mhsdesign commented:
For the following cases in the `WorkspaceCommandHandler` `ExpectedVersion::ANY()` is used while we could change that at least to `ExpectedVersion::STREAM_EXISTS()`? For most cases - because we issue multiple events - we dont want this check to ever fail, and it wont because the stream will continue to exist, even if someone deletes the workspace at the same time - that we have to prevent actually :D.

- handlePublishWorkspace
- handleRebaseWorkspace
- handlePublishIndividualNodesFromWorkspace
- handleDiscardIndividualNodesFromWorkspace
- discardWorkspace
- handleChangeBaseWorkspace
- handleDeleteWorkspace


Further i found a few missing `ExpectedVersion::ANY()` checks in the structure adjustments - but i think we dont have access to the content stream version here ...

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php:
```diff
@@ -148,7 +148,7 @@ private function handleCreateWorkspace(
                     $command->newContentStreamId,
                 )
             ),
-            ExpectedVersion::ANY()
```
this slight change has a big impact as now after deleting a workspace a new one with the exact same name cannot be recreated.

And because `\Neos\Neos\Domain\Service\WorkspaceService::getUniqueWorkspaceName` only prevents naming conflict with existing workspaces that are projected attempting to create a workspace with the same title via Ui fails.

steps:
1. create workspace 'MyTest'
2. delete workspace
3. repeat

<img width="442" alt="image" src="https://github.com/user-attachments/assets/1129a827-872d-4479-b520-bdce64276850" />


The error is as expected:

> Workspace could not be created.
> Expected version: -1 [no stream], actual version: 1

That means imo that `getUniqueWorkspaceName` needs to handle this case.

bwaidelich replied:
very good catch

kitsunet replied:
So this seems to hve no test then?

mhsdesign replied:
jup the the core's `DeleteWorkspace` is fully untested - i have that somewhere in my todo list and stumble upon that every once and then. Also the cores `CreateWorkspace` is only tested implicit but not for such cases.

mhsdesign replied:
how do we want to continue here? it seems like a good thing to prevent that only one workspace for that name can exist in an unpruned content repository?

bwaidelich replied:
As discussed: I reverted this part of the change.
To properly enforce hard constraints for workspace mutations, we'll need to project the version to the workspace read model. I'll try that in a follow-up, but this does not have to be changed for 9.0

mhsdesign replied:
Ill try to push a simple test that tests exactly that create delete create works!

bwaidelich committed:
Merge branch '9.0' into task/5058-use-proper-version-checks-for-contentStream-events

bwaidelich committed:
Revert checks for workspace creation

`ExpectedVersion::NO_STREAM()` would fail if a workspace was removed and another one _with the same name_ was created.

We'll need to project the workspace version to the workspace read model if we want to enforce hard constraints that are not prone to race conditions

mhsdesign committed:
TASK: Add test that deletion and recreation of a workspace should work using the same name

See discussion in https://github.com/neos/neos-development-collection/pull/5456#discussion_r1940837875
In the weekly we also agreed that a workspace must be recreateable if the name is free at the moment.

mhsdesign reviewed (approved):
Added test and assertion helpers ill use for other tests too :)
We need a full `DeleteWOrkspace` suite!
